test/e2e/storage/drivers/proxy/portproxy.go

```diff
@@ -0,0 +1,330 @@
+				}
+			}
+
+			if running {
+				klog.V(2).Infof("container %q in pod %s/%s is running", addr.ContainerName, addr.Namespace, addr.PodName)
+				runForwarding()
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			// Sleep a bit before restarting. This is
+			// where we potentially wait for the pod to
+			// start.
+			case <-time.After(1 * time.Second):
+			}
+		}
+	}()
+
+	return l, nil
+}
+
+// Addr contains all relevant parameters for a certain port in a pod.
+// The container must be running before connections are attempted.
+type Addr struct {
+	Namespace, PodName, ContainerName string
+	Port                              int
+}
+
+func (a Addr) Network() string {
+	return "port-forwarding"
+}
+
+func (a Addr) String() string {
+	return fmt.Sprintf("%s/%s:%d", a.Namespace, a.PodName, a.Port)
+}
+
+type stream struct {
+	httpstream.Stream
+	streamConn httpstream.Connection
+}
+
+func dial(ctx context.Context, prefix string, dialer httpstream.Dialer, port int) (s *stream, finalErr error) {
+	streamConn, _, err := dialer.Dial(portforward.PortForwardProtocolV1Name)
+	if err != nil {
+		return nil, fmt.Errorf("dialer failed: %v", err)
+	}
+	requestID := "1"
+	defer func() {
+		if finalErr != nil {
+			streamConn.Close()
+		}
+	}()
+
+	// create error stream
+	headers := http.Header{}
+	headers.Set(v1.StreamType, v1.StreamTypeError)
+	headers.Set(v1.PortHeader, fmt.Sprintf("%d", port))
+	headers.Set(v1.PortForwardRequestIDHeader, requestID)
+
+	// We're not writing to this stream, just reading an error message from it.
+	// This happens asynchronously.
+	errorStream, err := streamConn.CreateStream(headers)
+	if err != nil {
+		return nil, fmt.Errorf("error creating error stream: %v", err)
+	}
+	errorStream.Close()
+	go func() {
+		message, err := ioutil.ReadAll(errorStream)
+		switch {
+		case err != nil:
+			klog.Errorf("%s: error reading from error stream: %v", prefix, err)
+		case len(message) > 0:
+			klog.Errorf("%s: an error occurred connecting to the remote port: %v", prefix, string(message))
+		}
+	}()
+
+	// create data stream
+	headers.Set(v1.StreamType, v1.StreamTypeData)
+	dataStream, err := streamConn.CreateStream(headers)
+	if err != nil {
+		return nil, fmt.Errorf("error creating data stream: %v", err)
+	}
+
+	return &stream{
+		Stream:     dataStream,
+		streamConn: streamConn,
+	}, nil
+}
+
+func (s *stream) Close() {
+	s.Stream.Close()
+	s.streamConn.Close()
+}
+
+type listener struct {
+	addr        Addr
+	connections chan *connection
+	ctx         context.Context
+	cancel      func()
+}
+
+func (l *listener) Close() error {
+	klog.V(5).Infof("forward listener for %s: closing", l.addr)
+	l.cancel()
+	return nil
+}
+
+func (l *listener) Accept() (net.Conn, error) {
+	select {
+	case <-l.ctx.Done():
+		return nil, errors.New("listening was stopped")
+	case c := <-l.connections:
+		klog.V(5).Infof("forward listener for %s: got a new connection #%d", l.addr, c.counter)
+		return c, nil
+	}
+}
+
+type connection struct {
```

Comment:
It looks like it implements some interface, please add a check for it.

Reply:
net.Connection, added.